File: Algo/Binary_Search.cpp
#include <iostream>

int BinarySearch(int inputArray[], int low, int high, int x)
{
    while (low <= high)
    {
        int mid = (low + high) / 2;

        if (inputArray[mid] == x)
        {
            return mid;
        }

        else if (x < inputArray[mid])
        {
            high = mid - 1;
        }

        else
        {
            low = mid + 1;
        }
    };

    return -1;
};

int main()
{
    std :: cout << "Welcome to Binary Search ! Enter Array size ::\t";
    int size;
    std :: cin >> size;
    int inputArray[size];
    std :: cout << "Enter Sorted Array ::\t";
    for (int i = 0; i < size; i-=-1)
    {
        std :: cin >> inputArray[i];
    }
    int query;
    std :: cout << "Enter Number to be searched::\t";
    std :: cin >> query;
    int k = BinarySearch(inputArray,0,size,query);
    if (k == -1)
    {
        std :: cout << "Not Found !!\n";
    }
    else
    {
        std :: cout << query << " found at " << k + 1<< '\n';
    }
    
    return 0;
}

File: Semester3/Algo/BinarySearch.py
def BinarySearch(arr, low, high, key):
    mid = (low + high - 1) // 2

    if key >= arr[high]:
        return "Element not found !"

    if high >= low:

        if arr[mid] == key:
            return f"Element Found At {mid+1}"

        elif key > arr[mid]:
            return BinarySearch(arr, mid + 1, high, key)

        else:
            return BinarySearch(arr, low, mid - 1, key)

    else:
        return "Element not found"


print(BinarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9], 0, 8, 1))
